benchmarks/tpcc/src/main/java/com/google/cloud/pgadapter/tpcc/JavaClientBenchmarkRunner.java

```diff
@@ -0,0 +1,257 @@
+  void bindParams(Statement.Builder builder, Object[] params) throws SQLException {
+    for (int i = 0; i < params.length; i++) {
+      Object param = params[i];
+      if (param instanceof BigDecimal) {
+        builder.bind("p" + (i + 1)).to(Value.pgNumeric(((BigDecimal) param).toPlainString()));
+      } else if (param instanceof String) {
+        builder.bind("p" + (i + 1)).to((String) param);
+      } else if (param instanceof Integer) {
+        builder.bind("p" + (i + 1)).to((int) param);
+      } else if (param instanceof Long) {
+        builder.bind("p" + (i + 1)).to((long) param);
+      } else {
+        throw new SQLException(String.format("Unknown type for the parameter: %s", param));
+      }
+    }
+  }
+
+  Object[] paramQueryRow(String sql, Object[] params) throws SQLException {
+    ParametersInfo parametersInfo = convertPositionalParametersToNamedParameters(sql, '?', "$");
+    Statement.Builder builder = Statement.newBuilder(parametersInfo.sqlWithNamedParameters);
+    bindParams(builder, params);
+    Statement stmt = builder.build();
+
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    ResultSet resultSet = transactionThreadLocal.get().executeQuery(stmt);
+    Duration executionDuration = stopwatch.elapsed();
+    metrics.recordLatency(executionDuration.toMillis());
+
+    if (!resultSet.next()) {
+      throw new RowNotFoundException(String.format("No results found for: %s", sql));
+    }
+
+    Object[] result = new Object[resultSet.getColumnCount()];
+    for (int i = 0; i < result.length; i++) {
+      result[i] = getObject(resultSet.getValue(i));
+    }
+    return result;
+  }
+
+  void executeParamStatement(String sql, Object[] params) throws SQLException {
+    ParametersInfo parametersInfo = convertPositionalParametersToNamedParameters(sql, '?', "$");
+    Statement.Builder builder = Statement.newBuilder(parametersInfo.sqlWithNamedParameters);
+    bindParams(builder, params);
+    Statement stmt = builder.build();
+
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    transactionThreadLocal.get().executeUpdate(stmt);
+    Duration executionDuration = stopwatch.elapsed();
+    metrics.recordLatency(executionDuration.toMillis());
+  }
+
+  public static class ParametersInfo {
+    public final int numberOfParameters;
+    public final String sqlWithNamedParameters;
+
+    ParametersInfo(int numberOfParameters, String sqlWithNamedParameters) {
+      this.numberOfParameters = numberOfParameters;
+      this.sqlWithNamedParameters = sqlWithNamedParameters;
+    }
+  }
+
+  public ParametersInfo convertPositionalParametersToNamedParameters(
+      String sql, char paramChar, String queryParameterPrefix) {
+    final String namedParamPrefix = queryParameterPrefix;
+    StringBuilder named = new StringBuilder(sql.length() + countOccurrencesOf(paramChar, sql));
+    int index = 0;
+    int paramIndex = 1;
+    while (index < sql.length()) {
+      char c = sql.charAt(index);
+      if (c == paramChar) {
```

Comment:
This does not work if the SQL string contains for example comments or a string literal with a positional parameter.

Reply:
I looked at the original parser but that is way complicated for this benchmark. I think this simplified version works okay as all queries are hard-written and we know the generated strings.

Reply:
I tried to set `lock-scanned-ranges` to true which introduces some comments. This works fine. For `a string literal with a positional parameter`, I don't see why we would have such a case in the benchmark. This code just simply replace `?` to positional parameters with `$<num>` (later, also `@p<num>`).

Reply:
If you are certain that the benchmark does not have any SQL strings (and will not in the future) that might contain:
1. Comments that contain question marks.
2. String literals that contain question marks.

Then this should be fine.

But we should at least add a comment here to call that out clearly and explain that it is a deliberate choice, as a new developer who investigates this code in 2 years from now without any further context, might wonder how this ever worked.

Reply:
Good point. Added a comment.